numba/np/random/generator_core.py

```diff
@@ -0,0 +1,113 @@
+
+from llvmlite import ir
+from numba.core import cgutils, types
+from numba.core.extending import (intrinsic, make_attribute_wrapper, models,
+                                  overload, register_jitable,
+                                  register_model)
+from numba.core.imputils import Registry
+
+registry = Registry('generator_core')
+
+
+@register_model(types.NumPyRandomBitGeneratorType)
+class NumPyRngBitGeneratorModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ('parent', types.pyobject),
+            ('state_address', types.uintp),
+            ('state', types.uintp),
+            ('fnptr_next_uint64', types.uintp),
+            ('fnptr_next_uint32', types.uintp),
+            ('fnptr_next_double', types.uintp),
+            ('bit_generator', types.uintp),
+        ]
+        super(NumPyRngBitGeneratorModel, self).__init__(dmm, fe_type, members)
+
+
+_bit_gen_type = types.NumPyRandomBitGeneratorType('bit_generator')
+
+
+@register_model(types.NumPyRandomGeneratorType)
+class NumPyRandomGeneratorTypeModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ('bit_generator', _bit_gen_type),
+        ]
+        super(
+            NumPyRandomGeneratorTypeModel,
+            self).__init__(
+            dmm,
+            fe_type,
+            members)
+
+
+# The Generator instances have a bit_generator attr
+make_attribute_wrapper(
+    types.NumPyRandomGeneratorType,
+    'bit_generator',
+    'bit_generator')
+
+
+# Generate the overloads for "next_(some type)" functions
+def generate_next_binding(overloadable_function, return_type):
+    @intrinsic
+    def intrin_NumPyRandomBitGeneratorType_next_ty(tyctx, inst):
+        sig = return_type(inst)
+
+        def codegen(cgctx, builder, sig, llargs):
+            name = overloadable_function.__name__
+            struct_ptr = cgutils.create_struct_proxy(inst)(cgctx, builder,
+                                                           value=llargs[0])
+
+            # Get the 'state' and 'fnptr_next_(type)' members of the struct
+            state = struct_ptr.state
+            next_double_addr = getattr(struct_ptr, f'fnptr_{name}')
+
+            # LLVM IR types needed
+            ll_void_ptr_t = cgctx.get_value_type(types.voidptr)
+            ll_return_t = cgctx.get_value_type(return_type)
+            ll_uintp_t = cgctx.get_value_type(types.uintp)
+
+            # Convert the stored Generator function address to a pointer
+            next_fn_fnptr = builder.inttoptr(
+                next_double_addr, ll_void_ptr_t)
```

Comment:
That the function address is stored as an `int` is likely going to break caching as LLVM will bake it in as a constant.

Reply:
Do you have a specific case in which this should break ?

I tried doing this and it runs as expected (not sure if this is what you meant):

```python
import numpy as np
import numba

x = np.random.default_rng(1)
y = np.random.default_rng(1)

size = 2

@numba.njit(cache=True)
def do_stuff(gen):
    return gen.random(size)

print(do_stuff(y))
print(do_stuff(y))
print(do_stuff(y))
print(do_stuff(y))
# [0.51182162 0.9504637 ]
# [0.14415961 0.94864945]
# [0.31183145 0.42332645]
# [0.82770259 0.40919914]

print(x.random(8))
# [0.51182162 0.9504637  
#  0.14415961 0.94864945 
#  0.31183145 0.42332645
#  0.82770259 0.40919914]

```